File: src/test/kotlin/com/marcel/baumgardt/controller/ShowingControllerIntegrationTest.java
package com.marcel.baumgardt.controller;

import com.marcel.baumgardt.IntegrationTest;
import com.marcel.baumgardt.TestUtils;
import com.marcel.baumgardt.model.dto.request.UpdateShowingDatesRequest;
import com.marcel.baumgardt.model.dto.request.UpdateShowingPriceRequest;
import com.marcel.baumgardt.model.dto.response.ShowingDatesResponse;
import com.marcel.baumgardt.model.dto.response.UpdateShowingResponse;
import com.marcel.baumgardt.repository.ShowingRepository;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MvcResult;

import static com.marcel.baumgardt.TestConstants.*;
import static org.junit.jupiter.api.Assertions.assertEquals;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ShowingControllerIntegrationTest extends IntegrationTest {

    @Autowired
    private ShowingRepository showingRepository;

    @BeforeEach
    void setup() {
        //TODO this approach should be changed
        showingRepository.deleteAll();
        showingRepository.saveAll(TestUtils.Companion.createShowings());
    }

    @WithMockUser(username = "admin", authorities = {"ADMIN"})
    @Test
    void shouldSuccessfullyUpdatePriceOfShowingsAsAdmin() throws Exception {
        UpdateShowingPriceRequest request = TestUtils.Companion.createUpdateShowingPriceRequest();
        int affectedRows = getAffectedRows(request.getCinemaId(), request.getMovieId());
        UpdateShowingResponse expected = TestUtils.Companion.createSuccessfulUpdateShowingResponse(affectedRows);

        MvcResult mvcResult = performPutRequest(UPDATE_SHOWINGS_PRICE_URL, request);
        UpdateShowingResponse actual = getResponseObject(mvcResult, UpdateShowingResponse.class);

        assertEquals(actual, expected);
    }

    @WithMockUser(username = "user", authorities = {"USER"})
    @Test
    void shouldThrowForbiddenExceptionWhenUpdatePriceOfShowings() throws Exception {
        //TODO
    }

    @WithMockUser(username = "admin", authorities = {"ADMIN"})
    @Test
    void shouldSuccessfullyUpdateDatesOfShowings() throws Exception {

        UpdateShowingDatesRequest request = TestUtils.Companion.createSuccessfulUpdateShowingDatesRequest();

        UpdateShowingResponse expected = TestUtils.Companion.createNoEntitiesAffectedShowingResponse();

        MvcResult mvcResult = performPutRequest(UPDATE_SHOWINGS_DATES_URL, request);
        UpdateShowingResponse actual = getResponseObject(mvcResult, UpdateShowingResponse.class);

        assertEquals(actual, expected);
    }

    @WithMockUser(username = "admin", authorities = {"ADMIN"})
    @Test
    void shouldSuccessfullyClearOldUpdateDatesOfShowings() throws Exception {
        UpdateShowingDatesRequest request = TestUtils.Companion.createClearOldSuccessfulUpdateShowingDatesRequest();
        int affectedRows = getAffectedRows(request.getCinemaId(), request.getMovieId());
        UpdateShowingResponse expected = TestUtils.Companion.createSuccessfulUpdateShowingResponse(affectedRows);

        MvcResult mvcResult = performPutRequest(UPDATE_SHOWINGS_PRICE_URL, request);
        UpdateShowingResponse actual = getResponseObject(mvcResult, UpdateShowingResponse.class);

        assertEquals(actual, expected);
    }

    @WithMockUser(username = "user", authorities = {"USER"})
    @Test
    void shouldGerDatesOfShowingsByCinemaIdAndMovieId() throws Exception {
        ShowingDatesResponse expected = TestUtils.Companion.createShowingDatesResponse();

        MvcResult mvcResult = performGetRequest(GET_SHOWING_DATES_URL, SUCCESSFUL_CINEMA_ID, SUCCESSFUL_MOVIE_ID);
        ShowingDatesResponse actual = getResponseObject(mvcResult, ShowingDatesResponse.class);

        assertEquals(actual, expected);
    }

    private int getAffectedRows(Long cinemaId, Long movieId) {
        return showingRepository.countShowingsByCinemaIdAndMovieId(cinemaId, movieId);
    }

}

File: src/test/kotlin/com/marcel/baumgardt/controller/ShowingControllerIntegrationTest.kt
package com.marcel.baumgardt.controller

import com.marcel.baumgardt.IntegrationTest
import com.marcel.baumgardt.TestConstants.Companion.GET_SHOWING_DATES_URL
import com.marcel.baumgardt.TestConstants.Companion.SUCCESSFUL_CINEMA_ID
import com.marcel.baumgardt.TestConstants.Companion.SUCCESSFUL_MOVIE_ID
import com.marcel.baumgardt.TestConstants.Companion.UPDATE_SHOWINGS_DATES_URL
import com.marcel.baumgardt.TestConstants.Companion.UPDATE_SHOWINGS_PRICE_URL
import com.marcel.baumgardt.TestUtils.Companion.createClearOldSuccessfulUpdateShowingDatesRequest
import com.marcel.baumgardt.TestUtils.Companion.createShowingDatesResponse
import com.marcel.baumgardt.TestUtils.Companion.createSuccessfulUpdateShowingDatesRequest
import com.marcel.baumgardt.TestUtils.Companion.createSuccessfulUpdateShowingResponse
import com.marcel.baumgardt.TestUtils.Companion.createUpdateShowingPriceRequest
import com.marcel.baumgardt.model.dto.ShowingDatesResponse
import com.marcel.baumgardt.model.dto.UpdateShowingDatesRequest
import com.marcel.baumgardt.model.dto.UpdateShowingPriceRequest
import com.marcel.baumgardt.model.dto.UpdateShowingResponse
import com.marcel.baumgardt.repository.ShowingRepository
import lombok.RequiredArgsConstructor
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

@RequiredArgsConstructor
internal abstract class ShowingControllerIntegrationTest : IntegrationTest() {

    abstract val showingRepository: ShowingRepository

    @Test
    fun shouldSuccessfullyUpdatePriceOfShowings() {

        val request: UpdateShowingPriceRequest = createUpdateShowingPriceRequest()
        val affectedRows = getAffectedRows(request.cinemaId, request.movieId)
        val expected: UpdateShowingResponse = createSuccessfulUpdateShowingResponse(affectedRows)

        val mvcResult = performPutRequest(UPDATE_SHOWINGS_PRICE_URL, request)
        val actual: UpdateShowingResponse = getResponseObject(mvcResult, UpdateShowingResponse::class.java)

        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun shouldSuccessfullyUpdateDatesOfShowings() {

        val request: UpdateShowingDatesRequest = createSuccessfulUpdateShowingDatesRequest()
        val affectedRows = getAffectedRows(request.cinemaId, request.movieId)
        val expected: UpdateShowingResponse = createSuccessfulUpdateShowingResponse(affectedRows)

        val mvcResult = performPutRequest(UPDATE_SHOWINGS_DATES_URL, request)
        val actual: UpdateShowingResponse = getResponseObject(mvcResult, UpdateShowingResponse::class.java)

        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun shouldSuccessfullyClearOldUpdateDatesOfShowings() {
        val request: UpdateShowingDatesRequest = createClearOldSuccessfulUpdateShowingDatesRequest()
        val affectedRows = getAffectedRows(request.cinemaId, request.movieId)
        val expected: UpdateShowingResponse = createSuccessfulUpdateShowingResponse(affectedRows)

        val mvcResult = performPutRequest(UPDATE_SHOWINGS_PRICE_URL, request)
        val actual: UpdateShowingResponse = getResponseObject(mvcResult, UpdateShowingResponse::class.java)

        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun shouldGerDatesOfShowingsByCinemaIdAndMovieId() {
        val expected: ShowingDatesResponse = createShowingDatesResponse()

        val mvcResult = performGetRequest(GET_SHOWING_DATES_URL, SUCCESSFUL_CINEMA_ID, SUCCESSFUL_MOVIE_ID)
        val actual: ShowingDatesResponse = getResponseObject(mvcResult, ShowingDatesResponse::class.java)

        Assertions.assertEquals(expected, actual)
    }

    private fun getAffectedRows(cinemaId: Long, movieId: Long): Int {
        return showingRepository.countShowingsByCinemaIdAndMovieId(cinemaId, movieId)
    }
}